src/libmongoc/tests/test-mongoc-transactions.c

```diff
@@ -356,6 +356,47 @@ test_unknown_commit_result (void)
 }
 
 
+static void
+test_cursor_primary_read_pref (void *ctx)
+{
+   mongoc_client_t *client;
+   mongoc_client_session_t *session;
+   mongoc_collection_t *collection;
+   mongoc_cursor_t *cursor;
+   bson_t opts = BSON_INITIALIZER;
+   mongoc_read_prefs_t *read_prefs;
+   bson_error_t error;
+   bool r;
+
+   client = test_framework_client_new ();
+   collection = get_test_collection (client, "test_cursor_primary_read_pref");
+
+   session = mongoc_client_start_session (client, NULL, &error);
+   ASSERT_OR_PRINT (session, error);
+
+   r = mongoc_client_session_start_transaction (session, NULL, &error);
+   ASSERT_OR_PRINT (r, error);
+
+   r = mongoc_client_session_append (session, &opts, &error);
+   ASSERT_OR_PRINT (r, error);
+
+   read_prefs = mongoc_read_prefs_new (MONGOC_READ_PRIMARY);
+
+   cursor = mongoc_collection_find_with_opts (
+      collection, tmp_bson ("{}"), &opts, read_prefs);
+
+   bson_destroy (&opts);
+   mongoc_read_prefs_destroy (read_prefs);
+   mongoc_collection_destroy (collection);
+
+   ASSERT_OR_PRINT (!mongoc_cursor_error (cursor, &error), error);
```

Comment:
Let's iterate the cursor to completion first; this catches the current bug but I could easily imagine a new one that doesn't manifest until mongoc_cursor_next.